### PR TITLE
Make aiclib work with NSX 4.2

### PR DESCRIPTION
**Pull request by mpath**
Redirects and retries had to be contained within aiclib,
rather than in urllib3. Logins had to be redone on redirects.
Connection pools had to be recreated on redirect as well.

JIRA:NCP-1188

### CODE REVIEW

File: aiclib/common.py [hunk 1]
```diff
@@ -41,6 +40,7 @@ def is_booleanop(boolop):
 
 
 def genuri(*args):
+    logger.debug("[*] genuri(*args={%s}) called..." % (', '.join(args)))
```
Comment (Cerberus98) [inline aiclib/common.py]: Probably can remove this

File: aiclib/common.py [hunk 2]
```diff
@@ -22,8 +22,7 @@
 
 import calendar
 import logging
-import re
-
+from uuid import UUID
```
Comment (Cerberus98) [inline aiclib/common.py]: I'm not a big fan of pulling stuff from other modules into the local namespace

File: aiclib/nvp.py [hunk 3]
```diff
@@ -93,7 +91,7 @@ def gatewayservice(self, uuid=None):
         entity = nvpentity.GatewayService(self, uuid=uuidvalue)
         return entity
 
-    def transportzone(self, uuid=None):
+    def zone(self, uuid=None):
```
Comment (roaet) [inline aiclib/nvp.py]: Why change to zone but leave everything else as TransportZone? (curious)

**Comment by roaet** [comment]
Love your branch name

File: aiclib/common.py (hunk 1 above)
Comment (roaet) [inline aiclib/common.py]: +1

File: aiclib/common.py [hunk 4]
```diff
@@ -51,7 +51,10 @@ def dttounix(date_time):
...
 
 def isuuid(uuid):
-    regex = "%s-%s-%s-%s-%s" % ("[a-f0-9]{8}", "[a-f0-9]{4}",
-                                "[a-f0-9]{4}", "[a-f0-9]{4}",
-                                "[a-f0-9]{12}")
-    return re.match(regex, uuid)
+    if ' ' in uuid:
+        return False
+    try:
+        UUID(uuid, version=4)
```
Comment (Cerberus98) [inline aiclib/common.py]: Outside of this probably not belonging in this particular patch, we might as well just use the UUID constructor to decide.

File: aiclib/core.py [hunk 5]
```diff
@@ -30,6 +30,7 @@
     from urllib import urlencode
 
 import urllib3
+from urlparse import urlparse
```
Comment (Cerberus98) [inline aiclib/core.py]: Same. I'd rather import urlparse and call it as urlparse.urlparse everywhere

File: aiclib/core.py [hunk 6]
```diff
@@ -163,77 +162,100 @@ def headers(self):
...
-                url = url + '?' + params
-
+                new_url = url + '?' + params
             else:
                 if not is_body_prepared:
                     body = json.dumps(body)
-                open_kwargs['body'] = body
+                new_body = body
+        else:
+            if method in ("POST"):
```
Comment (Cerberus98) [inline aiclib/core.py]: What happens if it's a PUT and there's no body? We'll get a bad request from NSX but not know why possibly. I'd prefer we catch it here/nearby and raise a more useful error.

File: aiclib/nvp.py (hunk 3 above)
Comment (mpath) [inline aiclib/nvp.py]: There was a merge mess-up in 2013 where transportzone() was changed to just zone() in the tests but that change didn't make it to here. Also, test failures like that were ignored, so we've been missing it for a while.

**Comment by mpath** [comment]
Don't merge yet

**Comment by Cerberus98** [comment]
:+1:

**Comment by roaet** [comment]
looks fine. I can't merge here.

**Pull request merged**